arrows/vxl/hashed_image_classifier.cxx

```diff
@@ -0,0 +1,498 @@
+// This file is part of KWIVER, and is distributed under the
+// OSI-approved BSD 3-Clause License. See top-level LICENSE file or
+// https://github.com/Kitware/kwiver/blob/master/LICENSE for details.
+
+#include "hashed_image_classifier.h"
+
+#include <vital/range/iota.h>
+
+#include <vil/vil_plane.h>
+
+#include <cmath>
+
+namespace kwiver {
+
+namespace arrows {
+
+namespace vxl {
+
+// ----------------------------------------------------------------------------
+// Classify a chain of hashed feature images
+template < typename FeatureType, typename OutputType >
+void
+hashed_image_classifier< FeatureType, OutputType >
+::classify_images( input_image_t const& input_features,
+                   weight_image_t& output_image,
+                   weight_t const offset ) const
+{
+  feature_vector_t input_features_vector;
+  for( auto const i : vital::range::iota( input_features.nplanes() ) )
+  {
+    input_features_vector.push_back( vil_plane( input_features, i ) );
+  }
+  classify_images( input_features_vector,
+                   output_image,
+                   offset );
+}
+
+// ----------------------------------------------------------------------------
+// Classify a chain of hashed feature images
+template < typename FeatureType, typename OutputType >
+void
+hashed_image_classifier< FeatureType, OutputType >
+::classify_images( feature_vector_t const& input_features,
+                   weight_image_t& output_image,
+                   weight_t const offset ) const
+{
+  if( !model_->is_valid() )
+  {
+    LOG_ERROR( logger, "Internal classifier invalid" );
+  }
+
+  auto num_features = input_features.size();
+
+  if( num_features != feature_count() )
+  {
+    LOG_ERROR( logger, "Feature counts don't match, features: "
+               << num_features << ", feature_count(): " << feature_count() );
+  }
+
+  output_image.set_size(
+    input_features.at( 0 ).ni(), input_features.at( 0 ).nj() );
+  output_image.fill( offset );
+
+  std::vector< std::ptrdiff_t > sisteps( num_features );
+  std::vector< input_t const* > spixels( num_features );
+
+  for( auto const f : vital::range::iota( num_features ) )
+  {
+    sisteps[ f ] = input_features.at( f ).istep();
+  }
+
+  auto const distep = output_image.istep();
+
+  for( auto const j : vital::range::iota( output_image.nj() ) )
+  {
+    weight_t* dpixel = &output_image( 0, j );
+
+    for( auto const f : vital::range::iota( num_features ) )
+    {
+      spixels[ f ] = &input_features[ f ]( 0, j );
+    }
+
+    for( unsigned i = 0; i < output_image.ni(); ++i, dpixel += distep )
+    {
+      for( auto const f : vital::range::iota( num_features ) )
+      {
+        *dpixel += model_->feature_weights.at( f )[ *spixels[ f ] ];
```

Comment:
I'm not sure the bounds checking here (`at` vs. `operator[]`) is all that beneficial?